File: leetcode/1870.go
package leetcode

import (
	"math"
	"sort"
)

func minSpeedOnTime(dist []int, hour float64) int {
	h100 := int(math.Round(hour * 100))
	n := len(dist)
	if h100 <= (n-1)*100 {
		return -1
	}
	return 1 + sort.Search(1e7-1, func(v int) bool {
		v++
		h := 0
		for _, d := range dist[:n-1] {
			h += (d - 1 + v) / v
		}
		return (h*v+dist[n-1])*100 <= h100*v
	})
}

File: leetcode/1870.cpp
class Solution {
public:
    int minSpeedOnTime(vector<int>& dist, double hour) {
        int n = dist.size();
        long long t = llround(hour * 100);
        using ll = long long;
        if ((ll)(n - 1) * 100 >= t)
            return -1;
        int l = 1, r = 1e7;
        while (l < r) {
            int mid = l + (r - l) / 2;
            ll temp = 0;
            for (int i = 0; i < n - 1; i++) {
                temp += (long long)(dist[i] + mid - 1) / mid;
            }
            if ((temp * mid + dist[n - 1]) * 100 > t * mid) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return l;
    }
};